### PR TITLE
Allow testing for intentional leaks in test runner

### PR DESCRIPTION
**Pull request by Kelimion**
Adds `expect_leak_or_bad_free :: proc(t: ^T, client_test: proc(t: ^T), verifier: Memory_Verifier_Proc)`.

It sets up its own `Tracking_Allocator`, runs the `client_test`, and then calls the `verifier` procedure.
The verifier can then inspect the contents of the tracking allocator and call `testing.expect*` as sensible for the test in question.

Any allocations are then cleared so that the test runner doesn't itself complain about leaks.

Additionally, `ODIN_TEST_LOG_LEVEL_MEMORY` has been added as a define to set the severity of the test runner's memory tracker. You can use `-define:ODIN_TEST_LOG_LEVEL_MEMORY=error` to make tests fail rather than warn if leaks or bad frees have been found.

### CODE REVIEW

**Comment by Feoramund** [comment]
`-define:ODIN_TEST_LOG_LEVEL_MEMORY=error` does not actually cause a test to fail if it leaks though. Have I missed something here?

```odin
package main

import "core:testing"

@test
a :: proc(t: ^testing.T) {
	b := new(int)
}
```

`odin test . -define:ODIN_TEST_LOG_LEVEL_MEMORY=error -define:ODIN_TEST_SHORT_LOGS=true`

```
[INFO ] --- Starting test runner with 1 thread. Set with -define:ODIN_TEST_THREADS=n.
[INFO ] --- The random seed sent to every test is: 148072560220716. Set with -define:ODIN_TEST_RANDOM_SEED=n.
[INFO ] --- Memory tracking is enabled. Tests will log their memory usage if there's an issue.
[INFO ] --- < Final Mem/ Total Mem> <  Peak Mem> (#Free/Alloc) :: [package.test_name]
[ERROR] --- <        8B/        8B> <        8B> (    0/    1) :: main.a
        +++ leak         8B @ 0x750BCF000048 [memtest.odin:7:a()]
main  [|                       ]         1 :: [package done]

Finished 1 test in 193.756µs. The test was successful.
```

**Comment by Kelimion** [comment]
> `-define:ODIN_TEST_LOG_LEVEL_MEMORY=error` does not actually cause a test to fail if it leaks though. Have I missed something here?
> 
> ```odin
> package main
> 
> import "core:testing"
> 
> @test
> a :: proc(t: ^testing.T) {
> 	b := new(int)
> }
> ```
> 
> `odin test . -define:ODIN_TEST_LOG_LEVEL_MEMORY=error -define:ODIN_TEST_SHORT_LOGS=true`
> 
> ```
> [INFO ] --- Starting test runner with 1 thread. Set with -define:ODIN_TEST_THREADS=n.
> [INFO ] --- The random seed sent to every test is: 148072560220716. Set with -define:ODIN_TEST_RANDOM_SEED=n.
> [INFO ] --- Memory tracking is enabled. Tests will log their memory usage if there's an issue.
> [INFO ] --- < Final Mem/ Total Mem> <  Peak Mem> (#Free/Alloc) :: [package.test_name]
> [ERROR] --- <        8B/        8B> <        8B> (    0/    1) :: main.a
>         +++ leak         8B @ 0x750BCF000048 [memtest.odin:7:a()]
> main  [|                       ]         1 :: [package done]
> 
> Finished 1 test in 193.756µs. The test was successful.
> ```

Look at the return value. It now checks for bad frees + leaks if mem tracking is active.

File: core/testing/runner.odin [hunk 1]
```diff
@@ -63,6 +65,21 @@ get_log_level :: #force_inline proc() -> runtime.Logger_Level {
 	}
 }
 
+get_memory_log_level :: #force_inline proc() -> runtime.Logger_Level {
+	when ODIN_DEBUG {
+		// Always use .Debug in `-debug` mode.
+		return .Debug
+	} else {
```
Comment (Feoramund) [inline core/testing/runner.odin:72]: Is there a reason for this to default to Debug in `-debug` mode? The reason why general logging forces it to Debug is as a fallback to catch every possible log message, but I'm not sure why memory logging would do this, especially if there's an intent to catch leaks and errors by virtue of setting a higher log level.

Comment (Kelimion) [reply core/testing/runner.odin:72]: Not really, no. I just copy/pasted it with minimal edits.

**Pull request merged**